### PR TITLE
Transform intermediate function and complex struct support

### PR DESCRIPTION
**Pull request by keizer619**

### CODE REVIEW

File: modules/web/js/ballerina/components/transform-statement-decorator.jsx [hunk 1]
```diff
@@ -55,6 +50,18 @@ class TransformStatementDecorator extends React.Component {
 		};
 	}
 
+	componentDidMount() {
+        const { dragDropManager } = this.context;
+        dragDropManager.on('drag-start', this.startDropZones);
+        dragDropManager.on('drag-stop', this.stopDragZones);
+    }
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-statement-decorator.jsx]: There seems to be an indentation issue here.

File: modules/web/js/ballerina/components/transform-statement-decorator.jsx [hunk 2]
```diff
@@ -248,25 +280,98 @@ class TransformStatementDecorator extends React.Component {
...
+					let functionInvocationExpression = node.getChildren()[1].getChildren()[0];
+					let funPackage = this.context.renderingContext.packagedScopedEnvironemnt.getPackageByName(
+						functionInvocationExpression.getFullPackageName());
+					let func = funPackage.getFunctionDefinitionByName(functionInvocationExpression.getFunctionName());
+					this.mapper.addFunction(func, node, node.getParent().removeChild.bind(node.getParent()));
+				}
+			});
+	}
+
+    createConnection(statement){
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-statement-decorator.jsx]: There seems to be an indentation issue. Better to format the whole file.

File: modules/web/js/ballerina/components/transform-render.js [hunk 3]
```diff
@@ -35,858 +35,891 @@ var jsPlumb = jsPlumbLib.jsPlumb;
...
+}
+
+/**
+ * Disconnects the connection created.
+ *
+ * @param connection
+ */
+disconnect(connection) {
+    var self = this;
+    var propertyConnection = this.getConnectionObject(connection.getParameter("id"),
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-render.js]: We need to get rid of double quotes as per our eslint rules.

File: modules/web/js/ballerina/components/transform-render.js [hunk 4]
```diff
@@ -35,858 +35,891 @@ var jsPlumb = jsPlumbLib.jsPlumb;
...
+ */
+getConnectionObject(id, sourceId, targetId) {
+    var sourceName = this.getStructId(sourceId);
+    var targetName = this.getStructId(targetId);
+
+    var sourceRefObj;
+    var targetRefObj;
+
+    for (var i = 0; i < this.references.length; i++) {
+        if (this.references[i].name == sourceName) {
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-render.js]: Better to use ===

File: modules/web/js/ballerina/components/transform-render.js [hunk 5]
```diff
@@ -35,858 +35,891 @@ var jsPlumb = jsPlumbLib.jsPlumb;
...
-    getStructName(structId) {
-        return structId.split(this.viewIdSeperator)[0];
-    }
+/**
+ * Get the id of the struct from the propertyId
+ * @param propertyId Id of the property
+ * @returns {*}
+ */
+getStructId(propertyId) {
+    var id = propertyId.replace(this.jsTreePrefix + this.viewIdSeperator, "");
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-render.js]: We need to get rid of double quotes as per our eslint rules.

File: modules/web/js/ballerina/components/transform-render.js [hunk 6]
```diff
@@ -35,858 +35,891 @@ var jsPlumb = jsPlumbLib.jsPlumb;
...
-        return parts[parts.length - 1].split(this.nameTypeSeperator)[0];
-    }
+/**
+ * Get the type of the property from the property ID
+ * @param propertyId
+ * @returns {*}
+ */
+getPropertyType(propertyId) {
+    var parts = propertyId.split(this.idNameSeperator);
+    return parts[parts.length - 1].split(this.nameTypeSeperator)[1].replace('_anchor', "");
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-render.js]: We need to get rid of double quotes as per our eslint rules.

File: modules/web/js/ballerina/components/transform-render.js [hunk 7]
```diff
@@ -35,858 +35,891 @@ var jsPlumb = jsPlumbLib.jsPlumb;
...
-            connection.getOverlay("typeMapperDropdown").show();
-            var typeMapperId = '#typeMapperList' + this.viewIdSeperator + this.viewId;
-            var updatedTypeMapperId = "typeMapperList" + this.viewIdSeperator + connection.sourceId
+/**
+ * Populate the attribute properties stack from the propertyId
+ * @param propertyId
+ * @returns {Array}
+ */
+getPropertyNameStack(propertyId) {
+    var id = propertyId.replace(this.jsTreePrefix + this.viewIdSeperator, "");
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-render.js]: We need to get rid of double quotes as per our eslint rules.

File: modules/web/js/ballerina/components/transform-render.js [hunk 8]
```diff
@@ -35,858 +35,891 @@ var jsPlumb = jsPlumbLib.jsPlumb;
...
+processTypeMapperDropdown(info) {
+    var sourceType = this.getPropertyType(info.sourceId);
+    var targetType = this.getPropertyType(info.targetId);
+    var isValidTypes = sourceType == targetType;
+    var self = this;
+    if (!isValidTypes) {
+        var connection = info.connection;
+        connection.getOverlay("typeMapperDropdown").show();
+        var typeMapperId = '#typeMapperList' + this.viewIdSeperator + this.viewId;
+        var updatedTypeMapperId = "typeMapperList" + this.viewIdSeperator + connection.sourceId
```
Comment (hemikak) [inline modules/web/js/ballerina/components/transform-render.js]: We need to get rid of double quotes as per our eslint rules. I think theres a command to fix them automatically, @roonyh would know the command. We just need to give the path as a param.

Comment (roonyh) [inline modules/web/js/ballerina/components/transform-render.js]: `npm run lintfile -- <filename> --fix`  is the command

**Pull request merged**